thetis/utility.py

```diff
@@ -1004,3 +1007,162 @@ def solve(self, solve_w=True):
+    with open(filename, 'r') as f_in:
+        in_species_block = False
+        key_indentation = None
+        for line in f_in:
+            # Skip lines that do not contain a colon
+            if ':' not in line:
+                continue
+            # Discard comments
+            line = line.split('#')[0]
+            # Discard everything after the colon
+            line = line.split(':')[0]
+            # Find the line where the species block begins
+            if not in_species_block:
+                in_species_block = line == 'species'
+                continue
+            # Determine current key (which may or may not turn out to be a species key).
+            matches = re.findall(r'[^\s]+', line)
+            if len(matches) != 1:
+                continue
+            current_key = matches[0]
+            # Determine level of indentation
+            current_indentation = line.index(current_key)
+            # If not indented, the species block must have ended, and the
+            # rest of the file can be skipped.
+            if current_indentation == 0:
+                break
+            # If this is the first key within the species block, record the level of indentation
+            if key_indentation is None:
+                key_indentation = current_indentation
+            # Find all other keys with the same level of indentation as the first one
+            if current_indentation != key_indentation:
+                continue
+            # Append each key to the output list
+            species_keys.append(current_key)
+    return species_keys
+
+
+def read_tracer_from_yml(filename, function_space, lambdify_modules=None,
+                         preserve_order=True, append_dimension=False):
+    r"""
+    Constructs and returns an ordered dictionary that specifies an ADR model.
+
+    :arg filename: Path to a YAML file containing constant values, tracer
+        diffusivities and reaction terms.
+    :arg function_space: The Firedrake :class:`FunctionSpace` in which the
+        :class:`Function` for each tracer should reside.
+    :kwarg lambdify_modules: A string or dictionary to be passed as the
+        `modules` parameter of `sympy.lambdify`.
+    :kwarg preserve_order: If True, tracers will appear in the OrderedDict
+        in the same order as they appear in the YAML file.
+    :kwarg append_dimension: If `True`, a suffix will be appended to the
+        label of each species from the input file. The form of the suffix is
+        `_Nd`, where `N` is the dimensionality of the spatial domain in
+        which `function_space` resides.
+    """
+    # Load dictionary
+    model_dict = None
+    with open(filename, 'r') as f_in:
+        model_dict = yaml.safe_load(f_in)
+
+    key_order = None
+    if preserve_order:
+        # Ensure that parse_tracers_from_dict adds tracers in the same order
+        # as the corresponding species appear in the input file.
+        key_order = get_species_key_order(filename)
```

Comment:
I believe this part isn't strictly needed, because from Python 3.6 onwards an ordinary `dict` should preserve order by default. But I could be wrong.

Reply:
Firedrake requires Python 3.6 so I think we can drop `preserve_order`.

Reply:
Presumably that also means we can drop `get_species_key_order`?